README
Games-Domino

The  README  is  used  to  introduce the module and provide instructions on how to install the
module, any machine dependencies it may have (for example C compilers and installed libraries)
and any other information that should be provided before the module is installed.

A  README  file is required for CPAN modules since CPAN extracts the README file from a module
distribution  so  that  people  browsing the archive can use it to get an idea of the module's
uses.  It is usually a good idea to provide version information here so that people can decide
whether fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

Alternatively, to install with Module::Build, you can use the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Games::Domino

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Domino

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Games-Domino

    CPAN Ratings
        http://cpanratings.perl.org/d/Games-Domino

    Search CPAN
        http://search.cpan.org/dist/Games-Domino/

LICENSE AND COPYRIGHT

Copyright (C) 2012 Mohammad S Anwar

This  program  is  free  software; you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

Changes
Revision history for Games-Domino

0.05    Mon Aug 20 13:11:50 2012
        - Fixed the typo in the pod for "DESCRIPTION".

0.04    Sat Aug 18 21:33:29 2012
        - Updated the text for "DESCRIPTION".

0.03    Sat Aug 18 18:19:20 2012
        - Major changes to let the user play with the computer (interactive mode).
        - Fixed couple of bugs encountered while making changes.
        - Minor cosmetics changes.
        - Updated the sample game script (play-domino.pl) to reflect the changes in the core module.

0.02    Thu Jul 26 16:38:29 2012
        - Extended Games::Domino::Tile class to hold double property
          as well top/bottom. This is to support the vertical branch.
        - Added unit test around the new attributes.
        - Fixed few typo in the pod for Games::Domino::Tile class.

0.01    Thu Jul 26 04:00:00 2012
        - First version, released.

lib/Games/Domino.pm
package Games::Domino;

use 5.006;
use strict; use warnings;

use Carp;
use Mouse;
use Data::Dumper;
use List::Util qw(shuffle);
use Mouse::Util::TypeConstraints;
use overload ( '""'  => \&as_string );

use Games::Domino::Tile;
use Games::Domino::Player;

select(STDOUT);
$|=1;

$SIG{'INT'} = sub { print {*STDOUT} "\n\nCaught Interrupt (^C), Aborting the game.\n"; exit(1); };

=head1 NAME

Games::Domino - Interface to the Domino game.

=head1 VERSION

Version 0.05

=cut

our $VERSION = '0.05';

has 'stock'    => (is => 'rw', isa => 'ArrayRef[Games::Domino::Tile]');
has 'board'    => (is => 'rw', isa => 'ArrayRef[Games::Domino::Tile]');
has 'human'    => (is => 'rw', isa => 'Games::Domino::Player');
has 'computer' => (is => 'rw', isa => 'Games::Domino::Player');
has 'current'  => (is => 'rw', isa => 'Games::Domino::Player');
has 'board_l'  => (is => 'rw', isa => 'ZeroToSix');
has 'board_r'  => (is => 'rw', isa => 'ZeroToSix');
has 'cheat'    => (is => 'ro', isa => 'ZeroOrOne', default => 0);
has 'debug'    => (is => 'rw', isa => 'ZeroOrOne', default => 0);

=head1 DESCRIPTION

This is a very basic Domino game played by two players (Computer vs Human) at the  moment.This
is just an initial draft of Proof of Concept, also to get my head around the game which I have
never played in my life before. There is a cheat flag which makes tiles for "Computer" visible
to the other player "Human". Avoid this flag if possible. By default  the cheat flag is turned
off.There is a debug switch as well which is turned off by default.They are arranged like here
before we shuffle to start the the game.

    [0 | 0]
    [0 | 1] [1 | 1]
    [0 | 2] [1 | 2] [2 | 2]
    [0 | 3] [1 | 3] [2 | 3] [3 | 3]
    [0 | 4] [1 | 4] [2 | 4] [3 | 4] [4 | 4]
    [0 | 5] [1 | 5] [2 | 5] [3 | 5] [4 | 5] [5 | 5]
    [0 | 6] [1 | 6] [2 | 6] [3 | 6] [4 | 6] [5 | 6] [6 | 6]

    use strict; use warnings;
    use Games::Domino;

    my $domino = Games::Domino->new();

=cut

sub BUILD
{
    my $self  = shift;

    $self->{stock} = $self->_prepare();
    $self->{human} = Games::Domino::Player->new({ name => 'H', show => 1 });

    if ($self->cheat) {
        $self->{computer} = Games::Domino::Player->new({ name => 'C', show => 1 });
    } else {
        $self->{computer} = Games::Domino::Player->new({ name => 'C' });
    }

    $self->{human}->save(shift @{$self->{stock}})    for (1..7);
    $self->{computer}->save(shift @{$self->{stock}}) for (1..7);
    $self->{current} = $self->{computer};

    $self->_instructions;
    $self->show if $self->debug;
}

=head1 METHODS

=head2 play()

Pick a tile from the  current  player. If  no matching tile found then picks it from the stock
until it found one or the stock has only 2 tiles left at that time the game is over.

    use strict; use warnings;
    use Games::Domino;

    my $domino = Games::Domino->new();
    $domino->play;

=cut

sub play {
    my $self = shift;

    my $player = $self->{current};
    if ($player->name eq 'C') {
        my $tile = $player->pick($self->board_l, $self->board_r);
        if (defined $tile) {
            print {*STDOUT} "[C] [P]: $tile [S]\n" if $self->debug;
            $self->_save($tile);
            return $tile;
        }

        $tile = $self->_play($player->name);
        return $tile if defined $tile;

    } else {
        my $index;
        do {
            print {*STDOUT} "Pick your tile [" . $player->_available_indexes . "][B]? ";
            $index = <STDIN>;
            chomp($index) if defined $index;

            if (defined($index) && ($index =~ /B/i)) {
                my $tile = $self->_play($player->name);
                return $tile if defined $tile;
            }
        } until (defined($index) && $player->_validate_index($index) && $player->_validate_tile($index, $self->board_l, $self->board_r));

        my $tile = $player->_tile($index);
        print {*STDOUT} "[H] [P]: $tile [S]\n" if $self->debug;
        splice(@{$player->{bank}}, $index-1, 1);
        $self->_save($tile);
        return $tile;
    }
}

=head2 is_over()

Returns 1 or 0 depending whether the game is over or not. The game can be declared over in the
following circumstances:

=over 2

=item * Any one of the two players have used all his tiles.

=item * There are only two (2) tiles left in the bank.

=back

    use strict; use warnings;
    use Games::Domino;

    my $domino = Games::Domino->new();
    do { $domino->play; } until $domino->is_over;

=cut

sub is_over {
    my $self = shift;
    return 1 if ((scalar(@{$self->{stock}}) == 2)
                 ||
                 (scalar(@{$self->{human}->{bank}}) == 0)
                 ||
                 (scalar(@{$self->{computer}->{bank}}) == 0));
    return 0;
}

=head2 result()

Declares who is the winner against whom and by how much margin.

    use strict; use warnings;
    use Games::Domino;

    my $domino = Games::Domino->new();
    do { $domino->play; } until $domino->is_over;
    $domino->result;

=cut

sub result {
    my $self = shift;

    print {*STDOUT} "STOCK : $self\n";
    my $h = $self->{human}->value();
    my $c = $self->{computer}->value();
    if ($h == $c) {
        if (scalar(@{$self->{computer}->{bank}}) < scalar(@{$self->{human}->{bank}})) {
            print {*STDOUT} "WINNER: [C] [$c] against [H] [$h]\n";
        } else {
            print {*STDOUT} "WINNER: [H] [$h] against [C] [$c]\n";
        }
    } elsif ($h > $c) {
        print {*STDOUT} "WINNER: [C] [$c] against [H] [$h]\n";
    } else {
        print {*STDOUT} "WINNER: [H] [$h] against [C] [$c]\n";
    }
    $self->_line;
}

=head2 show()

Print the current tiles of Computer, Human and matched one.

    use strict; use warnings;
    use Games::Domino;

    my $domino = Games::Domino->new();
    do { $domino->play; $domino->show; } until $domino->is_over;

=cut

sub show {
    my $self = shift;
    $self->_line;
    print {*STDOUT} "[C]: " . $self->computer . "\n";
    print {*STDOUT} "[H]: " . $self->human    . "\n";
    print {*STDOUT} "[G]: " . $self->_board() . "\n";
    $self->_line;
}

=head2 as_string()

Returns all the unused tiles remained in the bank.

    use strict; use warnings;
    use Games::Domino;

    my $domino = Games::Domino->new();
    do { $domino->play; } until $domino->is_over;
    print "DOMINO : $domino\n\n";

=cut

sub as_string {
    my $self = shift;
    return '[EMPTY]' unless scalar(@{$self->{stock}});

    my $domino = '';
    foreach (@{$self->{stock}}) {
        $domino .= sprintf("%s==", $_);
    }
    $domino =~ s/[\=]+\s?$//;
    $domino =~ s/\s+$//;
    return $domino;
}

sub _instructions {
    my $self = shift;
    my $help = qq {
   _____                               _____                  _
  / ____|                          _ _|  __ \\                (_)
 | |  __  __ _ _ __ ___   ___  ___(_|_) |  | | ___  _ __ ___  _ _ __   ___
 | | |_ |/ _` | '_ ` _ \\ / _ \\/ __|   | |  | |/ _ \\| '_ ` _ \\| | '_ \\ / _ \\
 | |__| | (_| | | | | | |  __/\\__ \\_ _| |__| | (_) | | | | | | | | | | (_) \|
  \\_____|\\__,_|_| |_| |_|\\___||___(_|_)_____/ \\___/|_| |_| |_|_|_| |_|\\___/

Tiles are numbered left to right starting with 1. Symbols used in this game are:
    [C]: Code for the computer player
    [H]: Code for the human player
    [P]: Personal tile
    [B]: Tile picked from the bank
    [S]: Successfully found the matching tile
    [F]: Unable to find the matching tile
    [G]: All matched tiles so far

Example:

[C] [P]: [5 | 6] [S]
Computer picked the tile [5 | 6] from his own collection and successfully found the matching on board.

[H] [P]: [6 | 6] [S]
Human picked the tile [6 | 6] from his own collection and successfully found the matching on board.

[C] [B]: [2 | 6] [S]
Computer randomly picked the tile [2 | 6] from the bank and successfully found the matching on board.

[C] [B]: [3 | 4] [F]
Computer randomly picked the tile [3 | 4] from the bank and but failed to find the matching on board.

[H] [B]: [2 | 2] [S]
Human randomly picked the tile [2 | 2] from the bank and successfully found the matching on board.

[H] [B]: [3 | 6] [F]
Human randomly picked the tile [3 | 6] from the bank and but failed to find the matching on board.
};
    $self->_line;
    print {*STDOUT} $help,"\n";
    $self->_line;
}

sub _play {
    my $self   = shift;
    my $player = $self->{current};
    my $name   = $player->name;
    while (scalar(@{$self->{stock}}) > 2) {
        my $_tile = $self->_pick();
        $player->save($_tile);
        my $tile = $player->pick($self->board_l, $self->board_r);
        if (defined $tile) {
            print {*STDOUT} "[$name] [B]: $tile [S]\n" if $self->debug;
            $self->_save($tile);
            return $tile;
        } else {
            print {*STDOUT} "[$name] [B]: $_tile [F]\n" if $self->debug;
        }
    }
}

sub _save {
    my $self = shift;
    my $tile = shift;

    if (!defined($self->{board}) || (scalar(@{$self->{board}}) == 0)) {
        push @{$self->{board}}, $tile;
        $self->{board_l} = $tile->left;
        $self->{board_r} = $tile->right;
        $self->_next;
        return;
    }

    if ($self->{board_r} == $tile->left) {
        push @{$self->{board}}, $tile;
        $self->{board_r} = $tile->right;
        $self->_next;
        return;

    } elsif ($self->{board_r} == $tile->right) {
        my $L = $tile->left;
        my $R = $tile->right;
        $tile->right($L);
        $tile->left($R);
        push @{$self->{board}}, $tile;
        $self->{board_r} = $L;
        $self->_next;
        return;
    }

    if ($self->{board_l} == $tile->left) {
        my $L = $tile->left;
        my $R = $tile->right;
        $tile->right($L);
        $tile->left($R);
        unshift @{$self->{board}}, $tile;
        $self->{board_l} = $R;
        $self->_next;
        return;

    } elsif ($self->{board_l} == $tile->right) {
        unshift @{$self->{board}}, $tile;
        $self->{board_l} = $tile->left;
        $self->_next;
        return;
    }

    return;
}

sub _board {
    my $self = shift;

    my $board = '';
    foreach (@{$self->{board}}) {
        $board .= sprintf("%s==", $_);
    }
    $board =~ s/[\=]+\s?$//;
    $board =~ s/\s+$//;
    return $board;
}

sub _line {
    my $self = shift;
    print {*STDOUT} "="x76,"\n";
}

sub _pick {
    my $self = shift;
    return shift @{$self->{stock}};
}

sub _prepare {
    my $self  = shift;

    my $tiles = [];
    my $tile  = Games::Domino::Tile->new({ left => 0, right => 0, double => 1 });
    push @$tiles, $tile;
    foreach my $R (1..6) {
        my $L = 0;
        my $D = 0;
        while ($R >= $L) {
            ($R == $L)?($D = 1):($D = 0);
            push @$tiles, Games::Domino::Tile->new({ left => $L, right => $R, double => $D });
            $L++;
        }
    }

    $tiles = [shuffle @{$tiles}];
    return $tiles;
}

sub _next {
    my $self = shift;

    if ($self->{current}->name eq 'H') {
        $self->{current} = $self->{computer};
    } else {
        $self->{current} = $self->{human};
    }
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-games-domino at rt.cpan.org>, or  through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Domino>.I will be
notified,and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Domino

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Domino>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Domino>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Domino>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Domino/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Mohammad S Anwar.

This  program  is  free software;  you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 DISCLAIMER

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

__PACKAGE__->meta->make_immutable;
no Mouse;
no Mouse::Util::TypeConstraints;

1; # End of Games::Domino

lib/Games/Domino/Player.pm
package Games::Domino::Player;

use 5.006;
use strict; use warnings;

use Carp;
use Mouse;
use Mouse::Util::TypeConstraints;

use Data::Dumper;
use overload ( '""'  => \&as_string );

=head1 NAME

Games::Domino::Player - Represents the player of the Domino game.

=head1 VERSION

Version 0.05

=cut

our $VERSION = '0.05';

type 'HorC' => where { /^[H|C]$/i };

has 'name'  => (is => 'ro', isa => 'HorC', required => 1);
has 'bank'  => (is => 'rw', isa => 'ArrayRef[Games::Domino::Tile]');
has 'score' => (is => 'rw', isa => 'Int');
has 'show'  => (is => 'rw', isa => 'ZeroOrOne', default => 0);

=head1 DESCRIPTION

The Games::Domino::Player class is used by Games::Domino class internally.It shouldn't be used
directly.

=head1 METHODS

=head2 save()

Saves the given tile to the bank of the player.

    use strict; use warnings;
    use Games::Domino::Tile;
    use Games::Domino::Player;

    my $player = Games::Domino::Player->new({ name => 'H' });
    $player->save(Games::Domino::Tile->new({ left => 1, right => 4 }));

=cut

sub save {
    my $self = shift;
    my $tile = shift;

    croak("ERROR: Undefined tile found.\n") unless defined $tile;

    push @{$self->{bank}}, $tile;
}

=head2 value()

Returns the total value of all the tiles of the current player.

    use strict; use warnings;
    use Games::Domino::Tile;
    use Games::Domino::Player;

    my $player = Games::Domino::Player->new({ name => 'H' });
    $player->save(Games::Domino::Tile->new({ left => 1, right => 4 }));
    $player->save(Games::Domino::Tile->new({ left => 5, right => 3 }));
    print "The total value of the player is [" . $player->value . "]\n";

=cut

sub value {
    my $self  = shift;
    $self->{score} = 0;
    foreach (@{$self->{bank}}) {
        $self->{score} += $_->value;
    }
    return $self->{score};
}

=head2

Returns a matching tile for the given open ends. If no open ends found it then returns highest
value tile from the bank of the player.

    use strict; use warnings;
    use Games::Domino::Tile;
    use Games::Domino::Player;

    my $player = Games::Domino::Player->new({ name => 'H' });
    $player->save(Games::Domino::Tile->new({ left => 1, right => 4 }));
    $player->save(Games::Domino::Tile->new({ left => 5, right => 3 }));
    my $tile = $player->pick();
    print "Tile: $tile\n";

=cut

sub pick {
    my $self  = shift;
    my $left  = shift;
    my $right = shift;

    return $self->_pick($left, $right)
        if (defined($left) && defined($right));

    my $i    = 0;
    my $pos  = 0;
    my $max  = 0;
    my $tile = undef;

    foreach (@{$self->{bank}}) {
        if ($_->value > $max) {
            $max  = $_->value;
            $tile = $_;
            $pos  = $i;
        }
        $i++;
    }

    splice(@{$self->{bank}}, $pos, 1);
    return $tile;
}

=head2 as_string()

Returns the player object as string.This method is overloaded as string context.So if we print
the object then this method gets called. You can explictly call this method  as  well. Suppose
the player has 2 tiles then this return something like [1 | 4] == [5 | 3].

    use strict; use warnings;
    use Games::Domino::Tile;
    use Games::Domino::Player;

    my $player = Games::Domino::Player->new({ name => 'H' });
    $player->save(Games::Domino::Tile->new({ left => 1, right => 4 }));
    $player->save(Games::Domino::Tile->new({ left => 5, right => 3 }));
    print "Player: $player\n";

=cut

sub as_string {
    my $self = shift;
    my $bank = '';
    foreach (@{$self->{bank}}) {
        if ($self->show) {
            $bank .= sprintf("[%d | %d]==", $_->left, $_->right);
        } else {
            $bank .= sprintf("[x | x]==");
        }
    }
    $bank =~ s/[\=]+\s?$//;
    $bank =~ s/\s+$//;
    return $bank;
}

sub _pick {
    my $self  = shift;
    my $left  = shift;
    my $right = shift;

    my $i    = 0;
    my $pos  = 0;
    my $tile = undef;

    # Find all matching tiles.
    my $matched = {};
    foreach (@{$self->{bank}}) {
        my $L = $_->left;
        my $R = $_->right;
        if (($left =~ /$L|$R/) || ($right =~ /$L|$R/)) {
            $pos = $i;
            $tile = $_;
            $matched->{$i} = $tile;
        }
        $i++;
    }

    # Pick the maximum value tile among all the matched ones.
    my $pick = undef;
    my $max = 0;
    foreach (keys %{$matched}) {
        if ($matched->{$_}->value > $max) {
            $max = $matched->{$_}->value;
            $pick = { i => $_, t => $matched->{$_} };
        }
    }

    if (defined($pick)) {
        splice(@{$self->{bank}}, $pick->{i}, 1);
        return $pick->{t};
    }
    return;
}

sub _available_indexes {
    my $self = shift;

    return 1 if (scalar(@{$self->{bank}}) == 1);
    return "1..".scalar(@{$self->{bank}});
}

sub _validate_index {
    my $self  = shift;
    my $index = shift;

    return 0 unless (defined($index) && ($index =~ /^\d+$/));
    return 1 if ((scalar(@{$self->{bank}}) >= $index) && ($index >= 1));
    return 0;
}

sub _validate_tile {
    my $self  = shift;
    my $index = shift;
    my $left  = shift;
    my $right = shift;

    return 0 unless (defined($index) && ($index =~ /^\d+$/));
    return 1 unless (defined $left && defined $right);

    my $tile = $self->{bank}->[$index-1];
    my $L = $tile->left;
    my $R = $tile->right;

    return 1 if (($left =~ /$L|$R/) || ($right =~ /$L|$R/));
    return 0;
}

sub _tile {
    my $self  = shift;
    my $index = shift;
    return $self->{bank}->[$index-1];
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-games-domino at rt.cpan.org>,  or  through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Domino>. I will be
notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Domino::Player

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Domino>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Domino>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Domino>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Domino/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Mohammad S Anwar.

This  program  is  free software;  you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 DISCLAIMER

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

__PACKAGE__->meta->make_immutable;
no Mouse;
no Mouse::Util::TypeConstraints;

1; # End of Games::Domino::Player

lib/Games/Domino/Tile.pm
package Games::Domino::Tile;

use 5.006;
use strict; use warnings;

use Carp;
use Mouse;
use Mouse::Util::TypeConstraints;
use overload ( '""'  => \&as_string );

=head1 NAME

Games::Domino::Tile - Represents the tile of the Domino game.

=head1 VERSION

Version 0.05

=cut

our $VERSION = '0.05';

type 'ZeroOrOne' => where { /^[1|0]$/ };
type 'ZeroToSix' => where { /^[0-6]$/ };

has 'left'   => (is => 'rw', isa => 'ZeroToSix', required => 1);
has 'right'  => (is => 'rw', isa => 'ZeroToSix', required => 1);
has 'top'    => (is => 'rw', isa => 'ZeroToSix', required => 0);
has 'bottom' => (is => 'rw', isa => 'ZeroToSix', required => 0);
has 'double' => (is => 'ro', isa => 'ZeroOrOne', required => 1);

=head1 DESCRIPTION

The Games::Domino::Tile class is used by Games::Domino class internally. It  shouldn't be used
directly.

=cut

around BUILDARGS => sub
{
    my $orig  = shift;
    my $class = shift;

    unless (exists $_[0]->{double})
    {
        if (defined($_[0]->{left}) && defined($_[0]->{right}) && ($_[0]->{left} == $_[0]->{right}))
        {
            $_[0]->{double} = 1;
            $_[0]->{top} = $_[0]->{bottom} = $_[0]->{left};
        }
        else
        {
            $_[0]->{double} = 0;
        }
    }

    croak("ERROR: Invalid double attribute for the tile.\n")
        if (defined($_[0]->{left})
            &&
            defined($_[0]->{right})
            &&
            ( (($_[0]->{left} == $_[0]->{right})
               &&
               ($_[0]->{double} != 1))
              ||
              (($_[0]->{left} != $_[0]->{right})
               &&
               ($_[0]->{double} != 0)) )) ;

    if ($_[0]->{double} == 1)
    {
        $_[0]->{top} = $_[0]->{bottom} = $_[0]->{left};
    }

    return $class->$orig(@_);
};

=head1 METHODS

=head2 value()

Returns the value of the tile i.e. sum of left and right bips.

    use strict; use warnings;
    use Games::Domino::Tile;

    my $tile = Games::Domino::Tile->new({ left => 1, right => 4 });
    print "Value of the tile is [" . $tile->value . "].\n";

=cut

sub value
{
    my $self = shift;
    return ($self->{left} + $self->{right});
}

=head2 as_string()

Returns the tile object as string. This method is overloaded as string context. So if we print
the object then this method gets called. You can explictly call this method  as  well. Suppose
the tile has 3 left pips and 6 right pips then this would return it as [3 | 6].

    use strict; use warnings;
    use Games::Domino::Tile;

    my $tile = Games::Domino::Tile->new({ left => 1, right => 4 });
    print "The tile is $tile\n";
    # same as above
    print "The tile is " . $tile->as_string() . "\n";

=cut

sub as_string
{
    my $self = shift;
    return sprintf("[%d | %d]", $self->left, $self->right);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-games-domino at rt.cpan.org>,  or  through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Domino>. I will be
notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Domino::Tile

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Domino>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Domino>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Domino>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Domino/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Mohammad S Anwar.

This  program  is  free software;  you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 DISCLAIMER

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

__PACKAGE__->meta->make_immutable;
no Mouse;
no Mouse::Util::TypeConstraints;

1; # End of Games::Domino::Tile